### PR TITLE
Add enhanced event tracking with TTFT measurement and compact serialization.

### PR DESCRIPTION
**Pull request by lmcafee-nvidia**
# What does this PR do ?

  ## Summary

  This PR enhances the inference event tracking system with precise time-to-first-token (TTFT)
  measurement, per-token event tracking, and configurable serialization modes.

  **Key improvements:**
  - **Precise TTFT tracking**: Split request lifecycle into ADD_ENGINE and ADD_CONTEXT events to
  enable accurate TTFT measurement from request submission to first token generation
  - **Per-token event tracking**: New GENERATED_TOKEN events capture each token as it's generated with
   individual timestamps
  - **Compact serialization mode**: Add `--inference-dynamic-batching-track-generated-token-events`
  flag to control whether token events serialize as full dicts (with timestamps) or just token IDs,
  reducing overhead when timestamps aren't needed
  - **Comprehensive test suite**: Redesigned event tracking tests with 10 comprehensive test classes
  covering event ordering, timestamps, serialization modes, and edge cases

  **Changes:**
  - Event types: ADD_ENGINE, ADD_CONTEXT, GENERATED_TOKEN, FINISH
  - TTFT calculation: `first_GENERATED_TOKEN.timestamp - ADD_ENGINE.timestamp`
  - CLI flags: `--inference-dynamic-batching-track-generated-token-events` for verbose token event
  tracking
  - Test coverage: New integration tests for event timestamps and serialization modes

  **Bug fixes:**
  - Fixed FINISH event ordering in dynamic engine
  - Fixed DynamicInferenceRequest instantiation bug

  ## Test plan

  - [x] All existing inference functional tests pass
  - [x] New unit tests for event tracking (10 comprehensive test classes)
  - [x] Integration tests for event timestamps
  - [x] TTFT measurement validated in gpt_dynamic_inference.py output
  - [x] Compact vs. full serialization modes tested

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: examples/inference/gpt/gpt_dynamic_inference.py [hunk 1]
```diff
@@ -195,6 +373,17 @@ def _add_request():
...
+                add_engine_time = None
+                first_token_time = None
+                for event in finished_request.events:
+                    if event.type.name == "ADD_ENGINE":
+                        add_engine_time = event.timestamp
+                    elif event.type.name == "FIRST_TOKEN":
+                        first_token_time = event.timestamp
+                if add_engine_time is not None and first_token_time is not None:
+                    request.ttft = first_token_time - add_engine_time
+
```
Comment (tdene) [inline examples/inference/gpt/gpt_dynamic_inference.py]: Could we put this calculation & attribute directly in the engine instead? Right when we record the `FIRST_TOKEN` event, we can also assign this `ttft` attribute.

File: tests/unit_tests/inference/test_utils.py
Comment (tdene) [inline tests/unit_tests/inference/test_utils.py:1]: I'm not convinced this `TestPriority` is desirable. What are some arguments for it?

File: tests/unit_tests/inference/engines/test_dynamic_events.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        TEST_PRIORITY < TestPriority.CRITICAL, reason="Test priority not met"
+    )
+    def test_old_add_event_removed(self):
+        """Verify old ADD event type no longer exists."""
+        assert not hasattr(DynamicInferenceEventType, 'ADD')
+
+    @pytest.mark.skipif(
+        TEST_PRIORITY < TestPriority.IMPORTANT, reason="Test priority not met"
+    )
+    def test_all_event_types_exist(self):
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_events.py]: I don't think an explicit test for this is necessary; the rest of the tests should naturally cover this. But that's a style choice.

File: tests/unit_tests/inference/engines/test_dynamic_events.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    )
+    @pytest.mark.parametrize(
+        "event_type",
+        [
+            DynamicInferenceEventType.ADD_ENGINE,
+            DynamicInferenceEventType.ADD_CONTEXT,
+            DynamicInferenceEventType.FIRST_TOKEN,
+        ],
+    )
+    def test_event_creation(self, event_type):
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_events.py]: I don't think an explicit test for this is necessary.

File: tests/unit_tests/inference/engines/test_dynamic_events.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+        assert serialized['type'] == event_type.name
+        assert 'timestamp' in serialized
+
+        deserialized = DynamicInferenceEvent.deserialize(serialized)
+        assert deserialized.type == original.type
+        assert deserialized.timestamp == original.timestamp
+
+
+class TestRequestEventMethods:
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_events.py]: Most of these tests are redundant, because `test_event_timestamps_increasing` and `test_event_sequence` naturally cover all of them. No need for small tests if the big tests has equivalent coverage.

File: megatron/core/inference/inference_request.py [hunk 5]
```diff
@@ -225,32 +240,49 @@ def serialize(self) -> dict:
...
+
+                obj["payload"] = ContextErrorFactory.serialize(self.payload)
 
-            obj["payload"] = ContextErrorFactory.serialize(self.payload)
-        torch.cuda.nvtx.range_pop()
         return obj
 
     @classmethod
-    def deserialize(cls, obj: dict) -> "DynamicInferenceEvent":
+    def deserialize(cls, obj: dict | int) -> "DynamicInferenceEvent":
```
Comment (tdene) [inline megatron/core/inference/inference_request.py]: This is a general comment for the handling of `track_generated_token_events` throughout this diff.

Making the serialization/deserialization handle two different dtypes (dict / int) depending on the identity of the event and an argument is not a good pattern.

File: megatron/core/inference/inference_request.py [hunk 6]
```diff
@@ -225,32 +240,49 @@ def serialize(self) -> dict:
...
 
         Args:
-            obj (dict): Serialized event data.
+            obj: Serialized event data (dict for full event, int for compact GENERATED_TOKEN).
 
         Returns:
             (DynamicInferenceEvent) Deserialized event.
         """
+        # Handle compact GENERATED_TOKEN format
+        if isinstance(obj, int):
```
Comment (tdene) [inline megatron/core/inference/inference_request.py]: This is a really dangerous pattern. We should not assume that `int` means a specific type of event.

File: megatron/core/inference/inference_request.py [hunk 7]
```diff
@@ -288,6 +319,34 @@ def remaining_prompt_length(self):
 
     events: List[DynamicInferenceEvent] = field(default_factory=list)
 
+    @property
+    def generated_tokens(self) -> List[int]:
+        """Get the list of generated token IDs from GENERATED_TOKEN events.
```
Comment (tdene) [inline megatron/core/inference/inference_request.py]: The list comprehension in this method seems really costly. It will happen on every forward pass, and is a step back from just accessing the list.

In general, I don't understand why we would want to special-case `generated_tokens` as a property that is computed from events. I think it is fine to have both the events, which give detailed information, as well as the `generated_tokens` property which is efficient to append to and access.

File: megatron/core/inference/inference_request.py [hunk 8]
```diff
@@ -225,32 +240,49 @@ def serialize(self) -> dict:
...
+            if self.type == DynamicInferenceEventType.GENERATED_TOKEN:
+                # Token ID is already an int, no special serialization needed
+                obj["payload"] = self.payload
+            else:
+                from .contexts.dynamic_context import ContextErrorFactory  # avoid circular import.
+
+                obj["payload"] = ContextErrorFactory.serialize(self.payload)
 
-            obj["payload"] = ContextErrorFactory.serialize(self.payload)
-        torch.cuda.nvtx.range_pop()
```
Comment (tdene) [inline megatron/core/inference/inference_request.py:232]: Removing this line is a bug and will leave the range unclosed.

File: examples/inference/gpt/gpt_dynamic_inference.py [hunk 9]
```diff
@@ -195,6 +195,17 @@ def _add_request():
...
+                # Calculate TTFT = first GENERATED_TOKEN timestamp - ADD_ENGINE timestamp
+                add_engine_time = None
+                first_token_time = None
+                for event in finished_request.events:
+                    if event.type.name == "ADD_ENGINE":
+                        add_engine_time = event.timestamp
+                    elif event.type.name == "GENERATED_TOKEN" and first_token_time is None:
+                        first_token_time = event.timestamp
+                if add_engine_time is not None and first_token_time is not None:
+                    request.ttft = first_token_time - add_engine_time
```
Comment (tdene) [inline examples/inference/gpt/gpt_dynamic_inference.py]: Instead of calculating this after-the-fact, by iterating through the events list, can we instead calculate this directly inside the engine as soon as we get the first `GENERATED_TOKEN` event? It can be something like `request.ttft = None` at the beginning, then `request.ttft = (timestamp, )` upon `ADD_ENGINE`, then `request.ttft = timestamp - request.ttft[0]` upon the first `GENERATED_TOKEN`.

Or, better code than my suggestion above, we could have a `request.has_generated` boolean flag that we use to calculate TTFT.

File: megatron/training/arguments.py [hunk 10]
```diff
@@ -1594,6 +1594,11 @@ def _add_inference_args(parser):
                        help='Track paused request ids by adding \'paused\' events '
                        'to each request\'s event history. This has a very minor '
                        'impact on latency.')
+    group.add_argument('--inference-dynamic-batching-track-generated-token-events',
+                       action='store_true',
+                       help='Track generated token events with full timestamp info. '
+                       'When disabled, GENERATED_TOKEN events are serialized as '
+                       'just the token ID integer, reducing serialization overhead.')
```
Comment (tdene) [inline megatron/training/arguments.py]: If we move the TTFT calculation inside the engine, can we make this argument simply disable all `GENERATED_TOKEN` events altogether, instead of needing special-handling code to make the `GENERATED_TOKEN` events smaller and easier to serialize?

File: tests/unit_tests/inference/engines/test_dynamic_engine.py [hunk 11]
```diff
@@ -1091,27 +1091,89 @@ def test_events(self):
...
+            if request.status.name == 'COMPLETED':
+                # Completed requests should have ADD_ENGINE, ADD_CONTEXT, GENERATED_TOKEN(s), FINISH
+                assert event_types[0] in ('ADD_ENGINE', 'ERROR_TRANSIENT'), f"Unexpected first event: {event_types[0]}"
+                assert event_types[-1] == 'FINISH', f"Unexpected last event: {event_types[-1]}"
+                # Should have at least one GENERATED_TOKEN
+                assert 'GENERATED_TOKEN' in event_types, f"Missing GENERATED_TOKEN events"
+            elif request.status.name == 'FAILED':
+                # Failed requests should have ERROR_NONTRANSIENT, FAIL
+                assert 'ERROR_NONTRANSIENT' in event_types
+                assert event_types[-1] == 'FAIL'
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_engine.py]: The old test is more reliable. It asserted that all events happen in a specific, known, order.

File: tests/unit_tests/inference/engines/test_dynamic_engine.py [hunk 12]
```diff
@@ -1091,27 +1091,89 @@ def test_events(self):
...
-            ['ERROR_TRANSIENT', 'ADD', 'FINISH'],
-            ['ERROR_NONTRANSIENT', 'FAIL'],
-            ['ERROR_TRANSIENT', 'ADD', 'FINISH'],
-            ['ADD', 'FINISH'],
-            ['ERROR_TRANSIENT', 'ADD', 'FINISH'],
-            ['ERROR_TRANSIENT', 'ADD', 'FINISH'],
-            ['ERROR_TRANSIENT', 'ADD', 'FINISH'],
-        ]
-        result_event_types = [[e.type.name for e in r.events] for r in env.requests]
+        # Note: Each request generates 32 tokens, so there will be 32 GENERATED_TOKEN events per request
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_engine.py]: Is this `32 tokens` comment correct? Do some requests not finish early due to EOD tokens?

File: tests/unit_tests/inference/engines/test_dynamic_engine.py [hunk 13]
```diff
@@ -1091,27 +1091,89 @@ def test_events(self):
...
+        # For simplicity, we check that the sequence starts and ends correctly
+        for request in env.requests:
+            event_types = [e.type.name for e in request.events]
+            if request.status.name == 'COMPLETED':
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_engine.py]: In general, as a programming pattern, `request.status == Status.COMPLETED` is more reliable and future-facing.

File: tests/unit_tests/inference/engines/test_dynamic_engine.py [hunk 14]
```diff
@@ -1091,27 +1091,89 @@ def test_events(self):
...
+            num_requests=4,
+            max_prompt_length=16,
+            num_tokens_to_generate=num_tokens_to_generate,
+            context_buffer_size_gb=0.1,
+            num_gap_steps=0,
+        )
+
+        for request in env.requests:
+            if request.status != Status.COMPLETED:
+                continue
```
Comment (tdene) [inline tests/unit_tests/inference/engines/test_dynamic_engine.py]: Why do we skip COMPLETED requests altogether?

File: tests/unit_tests/inference/test_utils.py
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/test_utils.py:1]: I agree, and I removed it

File: tests/unit_tests/inference/engines/test_dynamic_events.py (hunk 2 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/engines/test_dynamic_events.py]: tests have been completely rewritten, please see latest commit

File: tests/unit_tests/inference/engines/test_dynamic_events.py (hunk 3 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/engines/test_dynamic_events.py]: tests have been completely rewritten, please see latest commit

File: tests/unit_tests/inference/engines/test_dynamic_events.py (hunk 4 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/engines/test_dynamic_events.py]: tests have been completely rewritten, please see latest commit

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 1 above)
Comment (lmcafee-nvidia) [reply examples/inference/gpt/gpt_dynamic_inference.py]: outdated, will respond to equivalent comment using GENERATED_TOKEN

File: megatron/core/inference/inference_request.py (hunk 7 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/inference_request.py]: I agree, I re-added the `generate_tokens` attribute

File: megatron/core/inference/inference_request.py (hunk 6 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/inference_request.py]: yeah you're right it's dangerous, I wasn't sure the best way to go about it. But I re-added the `generated_tokens` attribute, so this code is gone now

File: megatron/core/inference/inference_request.py (hunk 5 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/inference_request.py]: agreed and revamped the handling of `generated_tokens`

File: megatron/training/arguments.py (hunk 10 above)
Comment (lmcafee-nvidia) [reply megatron/training/arguments.py]: yep, this flag now controls whether `GENERATED_TOKEN` events are created at all

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 9 above)
Comment (lmcafee-nvidia) [reply examples/inference/gpt/gpt_dynamic_inference.py]: moved this to the engine's `post_process_requests()`. let me know what you think

File: tests/unit_tests/inference/engines/test_dynamic_engine.py (hunk 14 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/engines/test_dynamic_engine.py]: refactored this test and now it's gone

File: tests/unit_tests/inference/engines/test_dynamic_engine.py (hunk 13 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/engines/test_dynamic_engine.py]: you're right. claude is to blame. I'm just the messenger. half kidding, I just hadn't taken a close look yet

File: tests/unit_tests/inference/engines/test_dynamic_engine.py (hunk 11 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/engines/test_dynamic_engine.py]: reverted to old test

File: tests/unit_tests/inference/engines/test_dynamic_engine.py (hunk 12 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/engines/test_dynamic_engine.py]: reverted to the old test and removed this comment

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 9 above)
Comment (tdene) [reply examples/inference/gpt/gpt_dynamic_inference.py]: Looks great! Better than what I was suggesting.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 15]
```diff
@@ -172,6 +174,9 @@ def __init__(self, controller: TextGenerationController, context: DynamicInferen
         self.controller = controller
         self.context = context
         self.track_paused_request_events = inference_config.track_paused_request_events
+        self.track_generated_token_events = getattr(
+            inference_config, 'track_generated_token_events', False
+        )
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py]: nit: I don't believe the `getattr` part is needed, since it's being added to `inference_config.py`

**Comment by lmcafee-nvidia** [comment]
@shanmugamr1992 @Phlip79 , please review when you have time. thank you!

**Comment by ko3n1g** [comment]
🔄 Merge queue validation started!

You can track the progress here: https://github.com/NVIDIA/Megatron-LM/actions/runs/21958742266
